### PR TITLE
Immutable uploads time out if no writes are done for 30 minutes

### PR DESCRIPTION
**Pull request by itamarst**
Mostly useful for HTTP uploads.

Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3807

### CODE REVIEW

File: src/allmydata/storage/server.py [hunk 1]
```diff
@@ -14,12 +14,13 @@
 else:
     from typing import Dict
 
-import os, re, time
```
Comment (meejah) [inline src/allmydata/storage/server.py:17]: \O/

File: src/allmydata/storage/server.py [hunk 2]
```diff
@@ -135,6 +136,12 @@ def __init__(self, storedir, nodeid, reserved_space=0,
         # Canaries and disconnect markers for BucketWriters created via Foolscap:
         self._bucket_writer_disconnect_markers = {}  # type: Dict[BucketWriter,(IRemoteReference, object)]
 
+    def stopService(self):
+        # Cancel any in-progress uploads:
+        for bw in list(self._bucket_writers.values()):
```
Comment (meejah) [inline src/allmydata/storage/server.py:141]: probably don't need the `list()`?

File: src/allmydata/test/test_storage.py [hunk 3]
```diff
@@ -285,6 +285,66 @@ def count(self, name, delta=1):
...
+    def test_bucket_writes_delay_timeout(self):
+        """
+        So long as the ``BucketWriter`` receives writes, the the removal
+        timeout is put off.
+        """
+        incoming, final = self.make_workdir("test_bucket_writes_delay_timeout")
+        clock = Clock()
+        bw = BucketWriter(self, incoming, final, 200, self.make_lease(), clock)
+        # 20 minutes pass, getting close to the timeout...
+        clock.advance(29 * 60)
```
Comment (meejah) [inline src/allmydata/test/test_storage.py:320]: comment doesn't match code

**Review by meejah — APPROVED** [review]
LGTM, some inline comments

File: src/allmydata/storage/server.py (hunk 2 above)
Comment (itamarst) [reply src/allmydata/storage/server.py:141]: You do, when it gets disconnected it removes itself from `_bucket_writers` and then you're iterating over it as it gets mutated.

File: src/allmydata/test/test_storage.py (hunk 3 above)
Comment (itamarst) [reply src/allmydata/test/test_storage.py:320]: Fixed.

**Pull request merged**